DSI/globus_gridftp_server_iRODS.cpp

```diff
@@ -1642,12 +1665,12 @@ globus_l_gfs_iRODS_command(
                modAVUMetadataInp.arg4 = outChksum;
                modAVUMetadataInp.arg5 = arg5;
                rcModAVUMetadata(iRODS_handle->conn, &modAVUMetadataInp);
-
-           }
-
-           break;
+            }
+            break;
 
         default:
+            error_str = globus_common_create_string("iRODS: Command (%d) is not implemented.", cmd_info->command);
+            result = GlobusGFSErrorGeneric(error_str);
```

Comment:
What does `GlobusGFSErrorGeneric` do?

Reply:
It creates an error object which gets returned at the end.

Reply:
This got me to looking at the headers more.  There is a pound define for GlobusGFSNotImplemented.

I change the code to that and here is what is listed in the FTP output.  (Note that I commented out the rename code to get this to generate the error.)  How does this look?

```
ftp> rename abc.txt abc2
350 OK. Send RNTO with destination name.
500 GlobusError: v=1 c=NOT_IMPLEMETED
ftp> 
```

Reply:
I think I like that better - except... it has a typo in it...
`IMPLEMETED` -> `IMPLEMENTED`

what does the output look like with the original "iRODS: Command (%s) is not implemented."?

Reply:
I agree. It's pretty clean.

Reply:
Any idea what the `v=1` means? Is that the globus error code?

Reply:
I don't know what the v=1 is.  Trying to figure out how to use `GlobusGFSErrorNotImplemented` requires navigating through a chain of #defines.  I can't find any examples of its use.

I suspect the v=1 has something to do with the cause below.

```
#define GlobusGFSErrorNotImplemented()                                      \
        globus_error_put(GlobusGFSErrorObjNotImplemented(NULL))
#define GlobusGFSErrorObjNotImplemented(cause)                              \
        GlobusGFSErrorObj((cause), 500, "NOT_IMPLEMETED", NULL)
```
However, that cause is a globus_object_t pointer.  Here is the chain of defines and the definition of the globus_object_t.

```
typedef struct globus_object_s {
  const globus_object_type_t *   type;
  struct globus_object_s *       parent_object;
  void *                         instance_data;
  int                            ref_count;
} globus_object_t;

globus_gfs_ftp_response_error_construct(
    globus_module_descriptor_t *        base_source,
    globus_object_t *                   base_cause,
    int                                 response_code,
    const char                         *response_error_code,
    const char                         *fmt,
    ...);


#define GlobusGFSErrorObjFtpResponse(cause, code, response_error_code, ...) \
globus_gfs_ftp_response_error_construct(                                \
            NULL,                                                           \
            cause,                                                          \
            code,                                                           \
            response_error_code,                                            \
            __VA_ARGS__)
#define GlobusGFSErrorObj(cause, response_code, ...)                        \
    GlobusGFSErrorObjFtpResponse(cause, response_code, __VA_ARGS__)
```

Reply:
Did you ever investigate @trel's question here.

> what does the output look like with the original "iRODS: Command (%s) is not implemented."?

Also, consider asking the Globus devs about the macros, for example ... is usage of them recommended? Are there examples demonstrating proper usage?

Reply:
> Did you ever investigate @trel's question here.
> 
> > what does the output look like with the original "iRODS: Command (%s) is not implemented."?
> 
> Also, consider asking the Globus devs about the macros, for example ... is usage of them recommended? Are there examples demonstrating proper usage?

Here what the command could look like.  This is a little enhancement over what I had before.  These are just command numbers.  If it is a known one, I can translate it to the enumeration otherwise just give the numeric offset.

```
ftp> rename dir1 dir2
350 OK. Send RNTO with destination name.
500 iRODS: Command (GLOBUS_GFS_CMD_RNTO) is not implemented.
```

Reply:
I think that's better than the previous style.

Reply:
Agree.  I will make the change which will consist of the following:

1. Translate the various enumerations to the text string for the known enumerations in the header.
2. If I got a value that isn't in the enumeration, just write the integer value.

Reply:
As discussed offline, let's see if Globus can provide the conversion from enums to strings. That reduces the chance of getting an incorrect conversion.

If they aren't willing to do that, then the plan you've outlined sounds fine.

Reply:
I went ahead and implemented a map that translates the enumeration values to string values.  If the operation is not in the map, the operation number is printed.

Globus doesn't have a method for it.  We wouldn't get one soon even if they agreed to make the change and we would have to support older version anyway.

Their response was that "command not supported" is sufficient as the client knows what command was executed.  I actually kind of agree with this.  Maybe I was overthinking this.

I guess my thought was that if someone reported a "command not supported" it would be easier for us to troubleshoot if we were given the command description.

Thoughts?

Reply:
The user knows what they tried - but an admin looking at a log doesn't...

But yes - this is fine.   Let's get it in.